test/functional/sqlsrv/sqlsrv_batch_query.phpt

```diff
@@ -0,0 +1,204 @@
+        ++$e;
+    }
+}
+
+function checkColumnsAndRows($stmt, $cursor, $error)
+{
+    global $expectedCols, $expectedRows;
+    
+    $cols = sqlsrv_num_fields($stmt);
+    
+    if ($cols != $expectedCols) {
+        fatalError("Incorrect number of columns returned with $cursor cursor. Expected $expectedCols columns, got $cols columns\n");
+    }
+
+    $rows = sqlsrv_num_rows($stmt);
+
+    if ($cursor == 'buffered') {
+        if ($rows != $expectedRows) {
+            fatalError("Incorrect number of columns returned with buffered cursor. Expected $expectedRows rows, got $rows rows\n");
+        }
+    } else {
+        if ($rows !== false) {
+            fatalError("Expected sqlsrv_num_rows to return false with $cursor cursor, instead returned $rows rows\n");
+        } else {
+            checkErrors($error);
+        }
+    }
+}
+
+sqlsrv_configure('WarningsReturnAsErrors', 1);
+
+$conn = AE\connect();
+
+// Create and populate a table of integer types
+$tableName = 'batch_query_test';
+$columns = array(new AE\ColumnMeta('int', $colName[0]),
+                 new AE\ColumnMeta('tinyint', $colName[1]),
+                 new AE\ColumnMeta('smallint',$colName[2]),
+                 new AE\ColumnMeta('bigint', $colName[3]),
+                 new AE\ColumnMeta('bit', $colName[4]));
+                 
+$stmt = AE\createTable($conn, $tableName, $columns);
+if (!$stmt) {
+    fatalError("Failed to create table $tableName\n");
+}
+sqlsrv_free_stmt($stmt);
+
+$inputs = array();
+
+// Generate the inputs for insertRow()
+for ($i = 0; $i < $expectedRows; ++$i)
+{
+    $inputs[] = array();
+    for ($j = 0; $j < $expectedResultSets; ++$j)
+    {
+        $inputs[$i][$colName[$j]] = $data[$j][$i];
+    }
+}
+
+for ($i=0; $i < sizeof($inputs); ++$i) {
+    $stmt = AE\insertRow($conn, $tableName, $inputs[$i]);
+    sqlsrv_free_stmt($stmt);
+}
+
+$query = "SELECT c1_int from $tableName;
+          SELECT c2_tinyint from $tableName;
+          SELECT c3_smallint from $tableName;
+          SELECT c4_bigint from $tableName;
+          SELECT c5_bit from $tableName;";
+
+// Test the batch query with different cursor types
+for ($i = 0; $i < sizeof($cursors); ++$i)
+{
+    $cursor = $cursors[$i];
+    echo "Testing with $cursor cursor...\n";
+    
+    $stmt = sqlsrv_prepare($conn, $query, array(), array("Scrollable"=>$cursor));
+    if (!$stmt) { 
+        fatalError("Error preparing statement with $cursor cursor\n"); 
+    }
+
+    if (!sqlsrv_execute($stmt)) {
+        if ($cursor == 'forward' or $cursor == 'buffered') {
+            fatalError("Statement execution failed unexpectedly with a $cursor cursor\n");
+        } else {
+            checkErrors($noCursor);
+            continue;
+        }
+    }
+
+    $numResultSets = 0;
+
+    // Check the column and row count before and after running through
+    // each result set, because some cursor types may return the number
+    // of rows only after fetching all rows in the result set
+    do {
```

Comment:
I'm not sure I understand the logic of checking number of rows twice.  Here the cursor is either buffered or forward, and buffered cursor should always return the right number of rows, correct? I thought forward cursors always return false?

Reply:
On the PDO side, one of the server side cursors would return the number of rows once it was at the end of a result set, though I don't remember if it was a forward cursor. But I could easily imagine this is the kind of thing whose behaviour might change in the future, so I think we should leave it like this so that if the behaviour does change, it will trigger an error and force us to look at it again.